File: day1/part1.exs
defmodule Part1 do
  def solve(filename) do
    filename
    |> File.stream!()
    |> Stream.map(&parse_line/1)
    |> Enum.to_list()
    |> then(fn parsed_lines ->
      left = Enum.map(parsed_lines, &elem(&1, 0))
      right = Enum.map(parsed_lines, &elem(&1, 1))
      calculate_total_distance(left, right)
    end)
  end

  defp parse_line(line) do
    line
    |> String.trim()
    |> String.split("   ")
    |> Enum.map(&String.to_integer/1)
    |> List.to_tuple()
  end

  defp calculate_total_distance(left, right) do
    left = Enum.sort(left)
    right = Enum.sort(right)

    left
    |> Enum.zip(right)
    |> Enum.map(fn {l, r} -> abs(l - r) end)
    |> Enum.sum()
  end
end

"input.txt"
|> Part1.solve()
|> IO.puts()

File: day01/part1.lua
local function parse_line(line)
  local l, r = line:match("(%d+)%s+(%d+)")
  return tonumber(l), tonumber(r)
end

local function calculate_total_distance(left, right)
  table.sort(left)
  table.sort(right)

  local total = 0
  for i = 1, #left do
    total = total + math.abs(left[i] - right[i])
  end

  return total
end

local function solve(filename)
  local left = {}
  local right = {}

  for line in io.lines(filename) do
    local l, r = parse_line(line)
    table.insert(left, l)
    table.insert(right, r)
  end

  return calculate_total_distance(left, right)
end

print(solve("input.txt"))
